### ISSUE TITLE
wxLocale failure on MSW

### ISSUE THREAD

**Issue by paulcor** [wxWidgets/wxWidgets#22247]
With current master, `wxLocale("", "", setlocale(LC_ALL, nullptr), false)` causes an "invalid characters" assert in `wxLocaleIdent::FromTag()`, due to a space in the locale string, which is "English_United States.utf8". That string is the result of an earlier `setlocale(LC_ALL, ".utf8")`, which we use to enable utf8 support in the MSVC runtime.

This is on Windows 10 20H2 with MSVS 2022.

**Comment by utelle** [comment]
> With current master, `wxLocale("", "", setlocale(LC_ALL, nullptr), false)` causes an "invalid characters" assert in `wxLocaleIdent::FromTag()`, due to a space in the locale string, which is "English_United States.utf8". That string is the result of an earlier `setlocale(LC_ALL, ".utf8")`, which we use to enable utf8 support in the MSVC runtime.
> 
> This is on Windows 10 20H2 with MSVS 2022.

Argh, I really hate these Windows inconsistencies. If you look at the [setlocale documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170), you can get the impression that the return value of the locale name is in BCP47 syntax. That is, I would have expected a return value of `en-US.utf8` or `en_US.utf8`. Obviously this is not the case.

The question is whether we should add the space character to the list of allowed characters, or whether we should drop the check for invalid characters completely.

**Comment by utelle** [comment]
> I actually wanted to ask about this during the review, but forgot. The good news is that I think the answer to my unasked question about whether we should be using asserts in `FromTag()` at all is now clear: we shouldn't.

Hm, I assume that means all `wxASSERT_MSG` and `wxFAIL_MSG` statements should be removed. However, at least in debug mode I would find it useful to get some diagnostics why a tag is rejected. Can this be accomplished? If yes, how?

> If the input comes from outside of the program, it's not a programming error if it doesn't conform to the program expectations, i.e. it should be impossible for a user to trigger an assert in the program, but right now it can probably be done easily by setting `LC_ALL` to some invalid value under Unix.

Most likely you are right.

> So IMO there are 2 things to be done here:
>     1. `FromTag()` should be fixed to really accept "Windows" format, with spaces.

What exactly do you call "Windows format"? According to the [Locale Names documentation](https://docs.microsoft.com/en-us/windows/win32/intl/locale-names) a "locale name is based on the language tagging conventions of RFC 4646". And that format will be correctly interpreted by method `FromTag`. The combination (as described by @paulcor) returned by the C++ function `setlocale` does not correspond to this format (at least not always - no idea, on what ground the function chooses the format for the returned value) and will currently be rejected (even if we remove all asserts).

>     2. We should remove the other asserts from this function and try to do our best with whatever input we get.

Removing the asserts is easy, to retain debug diagnostics maybe less so. Method `FromTag` already tries to do its best to interpret the input in a reasonable fashion. However, if the input doesn't conform to any of the supported formats, it will be rejected (that is, return an empty `wxLocaleIdent`).

> @utelle Will you be able to do this please? This really must be fixed before 3.1.6.

When is the realease of 3.1.6 planned for?

Of course, I will try my best to fix this ASAP, but if method `FromTag` has to deal with the irregular locale tag format returned by `setlocale` on Windows it may take some more time. Most likely it would be easier to provide a helper function to convert the return value from `setlocale` to a valid locale tag.

@paulcor  From your description I get the impression that the call `wxLocale("", "", setlocale(LC_ALL, nullptr), false)` is done from from a user application? Is that correct? If yes, would it be an option to use a helper function (as mentioned above)?

**Comment by vadz** [comment]
> > I actually wanted to ask about this during the review, but forgot. The good news is that I think the answer to my unasked question about whether we should be using asserts in `FromTag()` at all is now clear: we shouldn't.
> 
> Hm, I assume that means all `wxASSERT_MSG` and `wxFAIL_MSG` statements should be removed.

Yes, the program shouldn't assert just because the user set some environment variable wrongly.

> However, at least in debug mode I would find it useful to get some diagnostics why a tag is rejected. Can this be accomplished? If yes, how?

The best way, IMO, would be to add unit tests for `FromTag()` that could be used to confirm that it handles the various inputs as expected. If you need something for debugging it, I'd use `wxLogTrace("uilocale", ...)` that could be enabled by setting `WXTRACE=uilocale`. But if we really consider it necessary to return this explanation to the program, then we need to add some `wxString *outError = NULL` parameter that would be filled with the message. I hope we can avoid doing this though.
 
> > So IMO there are 2 things to be done here:
> > 1. `FromTag()` should be fixed to really accept "Windows" format, with spaces.
> 
> What exactly do you call "Windows format"?

Sorry, I'm not being precise here, for me "Windows format" is "what Windows actually uses". Maybe I should have said "MSVC CRT format".

> According to the [Locale Names documentation](https://docs.microsoft.com/en-us/windows/win32/intl/locale-names) a "locale name is based on the language tagging conventions of RFC 4646". And that format will be correctly interpreted by method `FromTag`. The combination (as described by @paulcor) returned by the C++ function `setlocale` does not correspond to this format (at least not always - no idea, on what ground the function chooses the format for the returned value) and will currently be rejected (even if we remove all asserts).

This is definitely wrong. Either we need to accept this format too or MSW-specific code needs to be modified to use something different from `FromTag()`.

> > @utelle Will you be able to do this please? This really must be fixed before 3.1.6.
> 
> When is the realease of 3.1.6 planned for?

I want to release it on April 4, which ideally means finishing all the changes before the week-end.

> Of course, I will try my best to fix this ASAP, but if method `FromTag` has to deal with the irregular locale tag format returned by `setlocale` on Windows it may take some more time. Most likely it would be easier to provide a helper function to convert the return value from `setlocale` to a valid locale tag.

I think it's reasonable that people expect that passing a value returned by `setlocale()` as the `locale` argument of `wxLocale` ctor should work, even if I'm not sure that we explicitly guarantee that. So it's fine to have a helper function, but we need to call it in this ctor.
 
> @paulcor From your description I get the impression that the call `wxLocale("", "", setlocale(LC_ALL, nullptr), false)` is done from from a user application? Is that correct? If yes, would it be an option to use a helper function (as mentioned above)?

I don't think forcing changing the existing application code is worth it here. If we can continue supporting it (and I don't see why not), we should try to do it.

To summarize, my preferences would be:

1. Accept CRT locale strings in `FromTag()`: it already accepts different formats, what's one more.
2. Accept CRT locale strings in `wxLocale` ctor.
99. Document that such strings are not accepted any longer. But I'd really like to avoid this.

(the last item actually has the label of `99`, but GitHub markdown renderer insists on showing it as `3`).

**Comment by utelle** [comment]
> > Hm, I assume that means all `wxASSERT_MSG` and `wxFAIL_MSG` statements should be removed.
> 
> Yes, the program shouldn't assert just because the user set some environment variable wrongly.

Well, in my own applications I usually like to have diagnostics if something goes wrong.

> Sorry, I'm not being precise here, for me "Windows format" is "what Windows actually uses". Maybe I should have said "MSVC CRT format".

Unfortunately, M$ decided in its infinite wisdom to allow a CRT format that is not understood by the relevant Windows API functions. A problem is that up to now I have not found a Windows API function that allows to convert a CRT locale name (in the format `Language_Region.Codepage`, where "Language" and "Region" are given in natural English) to a locale name that can be used in Windows API functions. That is, the only chance to detect a matching locale name will be to search our language database for a matching entry. (Actually, the previous implementation of `wxLocale` fails to detect an entry in the language database, so that the language for the translation catalog is determined from the first 2 letters of the given locale name. I doubt that this will work for all languages - for example, for "German" one would get "ge", but required would be "de".)

What I also don't know is how the function `setlocale` decides which format is used for the returned locale name and whether this can be influenced somehow.

> [...] The combination (as described by @paulcor) returned by the C++ function `setlocale` does not correspond to this format (at least not always - no idea, on what ground the function chooses the format for the returned value) and will currently be rejected (even if we remove all asserts).
> 
> This is definitely wrong. Either we need to accept this format too or MSW-specific code needs to be modified to use something different from `FromTag()`.

I will try to integrate the handling for this format in method `FromTag`.

However, in Paul's example we have the locale name "English_United States.utf8". This name includes the codepage attribute "utf8", which is not part of the locale names of the Windows API. This might impose a problem.

> > When is the realease of 3.1.6 planned for?
> 
> I want to release it on April 4, which ideally means finishing all the changes before the week-end.

I will certainly try to complete the modifications before Saturday, but I can't guarantee that.

> I think it's reasonable that people expect that passing a value returned by `setlocale()` as the `locale` argument of `wxLocale` ctor should work, even if I'm not sure that we explicitly guarantee that. So it's fine to have a helper function, but we need to call it in this ctor.

Maybe the situation is not as bad as I feared. I just checked the implementation of `wxLocale::Init`. The locale name parameter will be used unchanged in the call to `setlocale`.  So, only the call `wxUILocale::UseLocaleName` must not fail and should initialize a `wxUILocale` instance based on language and region (the codepage part should not matter here). 

> To summarize, my preferences would be:
> 
> 1. Accept CRT locale strings in `FromTag()`: it already accepts different formats, what's one more.
> 2. Accept CRT locale strings in `wxLocale` ctor.
> 3. Document that such strings are not accepted any longer. But I'd really like to avoid this.

I'll try to implement variant 1.

**Comment by vadz** [comment]
FWIW the string returned by `setlocale()` seems to conform to the format described [here](https://docs.microsoft.com/en-us/cpp/c-runtime-library/locale-names-languages-and-country-region-strings), taking into account that "country names" refers to the countries [supported by Win32](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-lcid/a9eac961-e77d-41a6-90a5-ce1a8b0cdb9c) which includes "United States". In particular, ".utf8" is part of this (CRT) format.

**Comment by paulcor** [comment]
> @paulcor From your description I get the impression that the call wxLocale("", "", setlocale(LC_ALL, nullptr), false) is done from a user application?

For the record, the actual usage is something like:
```
class App: public wxApp {
    wxLocale m_locale;
    App() : m_locale("", "", setlocale(LC_ALL, nullptr), false) { }
    ...
};
```

Somewhat ironically, it looks like `wxUILocale::UseDefault()` is probably what we actually want, so I'll likely switch to that after 3.1.6 is available.

**Comment by utelle** [comment]
Actually, the problem was not in method `wxLocaleIdent::FromTag`, but in method `wxLocale::FindLanguageInfo` aka `wxUILocale::FindLanguageInfo`. The method description reads:
> This function may be used to find the language description structure for the given locale, specified either as a two letter ISO language code (for example, "pt"), a language code followed by the country code ("pt_BR") or a full, human readable, language description ("Portuguese-Brazil")

The variant where the locale is given as a full, human readable, language description never works (and never worked in previous releases), unless only the language is given. `setlocale` would return "Portuguese_Brazil" (please note the underscore). In method `wxLocale::FindLanguageInfo` the full string is used for comparison with the description in a language entry of the database, and therefore neither the entry with description "Portuguese (Brazil)" nor that with description "Portuguese" is found.

In the meantime I adjusted method `wxLocale::FindLanguageInfo` to properly search the language database, when the locale is given in Windows CRT format (i.e. "language[_region][.codeset]"). Thereafter the `wxLocale` ctor sets the locale via the language database entry, and method `wxUILocale::FromTag` is not called at all.

A potential problem may be that for a locale name with codepage (i.e. "English_United States.utf8") the function `setlocale` will actually be called now with locale "en_US" - note that the codeset attribute is gone. In prior releases `setlocale` was called with the full locale string, because method `wxLocale::FindLanguageInfo` failed due to the wrong implementation.

IMHO correcting the behaviour of `wxLocale::FindLanguageInfo` is the right thing to do. The question is how `wxLocale::Init` should behave, if the locale name contains a codepage attribute (as in @paulcor 's use case).

**Comment by utelle** [comment]
> > The point is that this page describes the format that is understood by CRT function `setlocale` (and friends), but `wxUILocale` does not use `setlocale`.

Actually, my remark about `wxUILocale` not using `setlocale` is not completely correct: under Unix systems (wxGTK) `setlocale` is used, but typically UTF-8 codepage is selected automatically.

> Yes, but this is a real problem because traditionally `wxLocale` was described as "portable wrapper around `setlocale()`" (hence its name...), so it was reasonable for people to expect passing strings in `setlocale()` format to it to work. So maybe we should actually put the workaround only in `wxLocale` itself -- as we don't need to make any such claims for `wxUILocale`.

To be on the safe side it will be good to enable method `wxLocaleIdent::FromTag` to handle Windows CRT locale names. I have already applied the necessary changes.

> > In the meantime I adjusted method wxLocale::FindLanguageInfo to properly search the language database, when the locale is given in Windows CRT format (i.e. "language[_region][.codeset]"). Thereafter the wxLocale ctor sets the locale via the language database entry, and method wxUILocale::FromTag is not called at all.
>
> So I think this is indeed the right thing to do, thanks!

Yes, but there is a **but**, because this changes the current behaviour in method `wxLocale::Init`. We have to discuss how the code should work (see below).

> > A potential problem may be that for a locale name with codepage (i.e. "English_United States.utf8") the function setlocale will actually be called now with locale "en_US" - note that the codeset attribute is gone.
>
> But this is still a problem :-( We need to extract and preserve it separately because people who want to use UTF-8 definitely want to use UTF-8 and not CP1252.

Yes, I fully agree. Therefore we need to adjust the behaviour of the `wxLocale` ctor resp the method `wxLocale::Init`. The declarations are as follows:
```C++
wxLocale (const wxString &name, const wxString &shortName=wxEmptyString, const wxString &locale=wxEmptyString, bool bLoadDefault=true);
bool Init (const wxString &name, const wxString &shortName=wxEmptyString, const wxString &locale=wxEmptyString, bool bLoadDefault=true);
```
- The parameter `name` is only descriptive and not used for anything else.
- The parameter `shortName` is a canonical name, if specified.
- The parameter `locale` is the locale name to be used in calling `setlocale`. If `locale` is empty, `shortName` will be used for this purpose.

Now, the relevant code in method `wxLocale::Init`:
```C++
    // change current locale (default: same as long name)
    wxString szLocale(locale);
    if ( szLocale.empty() )
    {
        // the argument to setlocale()
        szLocale = shortName;
        wxCHECK_MSG( !szLocale.empty(), false, wxS("no locale to set in wxLocale::Init()") );
    }

    if ( const wxLanguageInfo* langInfo = FindLanguageInfo(szLocale) )
    {
        // Prefer to use Init(wxLanguage) overload if possible as it will
        // correctly set our m_language and also set the locale correctly under
        // MSW, where just calling wxSetLocale() as we do below is not enough.
        //
        // However don't do it if the parameters are incompatible with this
        // language, e.g. if we are called with something like ("French", "de")
        // to use French locale but German translations: this seems unlikely to
        // happen but, in principle, it could.
        if ( langInfo->CanonicalName.StartsWith(shortName) )
        {
            return Init(langInfo->Language,
                        bLoadDefault ? wxLOCALE_LOAD_DEFAULT : 0);
        }
    }
```
I propose the following changes:

1. If the search in the language database was successful, call the language id based variant of the `Init` method only, if the `locale` parameter was empty, that is, `if ( locale.empty() && langInfo->CanonicalName.StartsWith(shortName) )`. Interestingly, in Paul's use case `shortName` is empty, but nevertheless `StartsWith` returns `true`.
2. If the search in the language database was successful and the `locale` parameter was **not** empty, set member variables `m_strLocale` and `m_language` as if method `Init` had been called.
3. If the search in the language database was successful, use the canonical name for instantiating the `wxUILocale`; otherwise use the `locale` parameter.

This way it will be respected, if a user explicitly requests a specific codepage using a Windows CRT locale name.

If this sounds reasonable, I will implement it that way tomorrow and will then make a PR.

**Comment by vadz** [comment]
Just to confirm, do I understand correctly that `wxLocale("", "", "French_France.utf8")` doesn't work now and didn't work before neither, because it doesn't (and didn't) switch to French locale at Windows level? I.e. that specifying `locale` parameter in the "Country_Language" form never actually worked under MSW except if the system was already using this combination?

If so, I'm not sure what should we be trying to do here: make it work exactly as before for compatibility or make it work as would be expected, i.e. call `setlocale()` using `locale` if it's non-empty but still set the locale at MSW level.

The latter would mean calling `Init(langInfo->Language)` if and only if `locale` is "compatible" with `shortName`, which is definitely more involved than using `wxString::StartsWith()`, but probably could be done by using `FindLanguageInfo()` for both of them and comparing the results?

**Comment by utelle** [comment]
> Just to confirm, do I understand correctly that `wxLocale("", "", "French_France.utf8")` doesn't work now and didn't work before neither, because it doesn't (and didn't) switch to French locale at Windows level?

If the locale name contains invalid characters like in Paul's use case ("English_United States.utf8"), then it doesn't work currently, because you get an assertion. For your example ("French_France.utf8") the instantiation of `wxUILocale` would fail, because `FromTag` does not yet handle Windows CRT locale names, but `setlocale` would be called successfully. However, only because `FindLanguageInfo` fails.

In the previous implementation (without `wxUILocale`) it worked, because `FindLanguageInfo` failed and `setlocale` was then called with the given locale name. 

> I.e. that specifying `locale` parameter in the "Country_Language" form never actually worked under MSW except if the system was already using this combination?

No, it worked, because `FindLanguageInfo` failed and `setlocale` accepts the "Language_Country" form under MSW. That is, it worked, because the implementation of `FindLanguageInfo` is flawed. According to the description `FindLanguageInfo` should handle the "Language_Country" form, but up to now it never did.

> If so, I'm not sure what should we be trying to do here: make it work exactly as before for compatibility or make it work as would be expected, i.e. call `setlocale()` using `locale` if it's non-empty but still set the locale at MSW level.

First, method `FindLanguageInfo` should be corrected, so that it works as described. However, this requires changes in `wxLocale::Init`, so that not the locale id based variant of `Init` is called (calling `setlocale` with the possibly wrong locale name), but instead `setlocale` is called with the given explicit locale name.

The user rightly expects that `wxLocale` respects an explicitly specified locale like "French_France.utf8". This is what my proposal will accomplish.

> The latter would mean calling `Init(langInfo->Language)` if and only if `locale` is "compatible" with `shortName`, which is definitely more involved than using `wxString::StartsWith()`, but probably could be done by using `FindLanguageInfo()` for both of them and comparing the results?

The idea is to never call `Init(langInfo->Language)` if `locale` is not empty, so that `setlocale` is definitely called with the `locale` value specified by the user (assuming that the user knows what he is doing). Nevertheless, `FindLanguageInfo()` should be used to find a matching language database entry. If an entry could be found, it should be used for instantiating the `wxUILocale` (for which we don't need the codepage information, even if it was part of the `locale` parameter). In this way the CRT functions will use the `setlocale` settings (as requested), and the wx classes will use the `wxUILocale` settings (as intended).

If a new application uses only `wxUILocale`(as it should, if I understood you correctly), then the wx classes will use the `wxUILocale` settings, while the CRT functions will stick to the standard C locale. At least under MSW and OSX, where `wxUILocale` doesn't call `setlocale`.

**Comment by vadz** [comment]
> > Just to confirm, do I understand correctly that `wxLocale("", "", "French_France.utf8")` doesn't work now and didn't work before neither, because it doesn't (and didn't) switch to French locale at Windows level?
> 
> If the locale name contains invalid characters like in Paul's use case ("English_United States.utf8"), then it doesn't work currently, because you get an assertion.

Sorry, I understand this. But I think that it didn't work before neither.

> In the previous implementation (without `wxUILocale`) it worked, because `FindLanguageInfo` failed and `setlocale` was then called with the given locale name.

Yes, but this didn't mean that it really worked because the system would still be using English and not French at MSW level. E.g. message boxes would be using Yes/No and not Oui/Non as would be expected.
 
> The idea is to never call `Init(langInfo->Language)` if `locale` is not empty, so that `setlocale` is definitely called with the `locale` value specified by the user (assuming that the user knows what he is doing). Nevertheless, `FindLanguageInfo()` should be used to find a matching language database entry. If an entry could be found, it should be used for instantiating the `wxUILocale` (for which we don't need the codepage information, even if it was part of the `locale` parameter). In this way the CRT functions will use the `setlocale` settings (as requested), and the wx classes will use the `wxUILocale` settings (as intended).

OK, this would be fix the problem above, i.e. go even beyond restoring compatibility as it will now work better than before.

> If a new application uses only `wxUILocale`(as it should, if I understood you correctly)

Yes, portable applications should only use it because what `wxLocale` does just can't be done under macOS.

**Comment by utelle** [comment]
> [...] but this didn't mean that it really worked because the system would still be using English and not French at MSW level. E.g. message boxes would be using Yes/No and not Oui/Non as would be expected.

You are right - in that respect it didn't work.

> OK, this would be fix the problem above, i.e. go even beyond restoring compatibility as it will now work better than before.

As said I will try to finish the implementation tomorrow and to make a PR.
